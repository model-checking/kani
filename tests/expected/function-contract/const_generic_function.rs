// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

//! Check that Kani contract can be applied to a constant generic function.
//! <https://github.com/model-checking/kani/issues/3667>

struct Foo<T> {
    ptr: *const T,
}

impl<T: Sized> Foo<T> {
    #[kani::requires(true)]
    pub const unsafe fn bar(self, v: T)
    where
        T: Sized,
    {
        unsafe { core::ptr::write(self.ptr as *mut T, v) };
    }
}

#[kani::proof_for_contract(Foo::bar)]
fn check_const_generic_function() {
    let x: u8 = kani::any();
    let foo: Foo<u8> = Foo { ptr: &x };
    unsafe { foo.bar(kani::any::<u8>()) };
}
