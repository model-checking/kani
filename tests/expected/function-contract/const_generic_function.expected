 ** 1 of
Failed Checks: Check that *dst is assignable
